opm/simulators/wells/WellInterface_impl.hpp

```diff
@@ -1624,29 +1625,65 @@ namespace Opm
             }
         }
 
-        // Calculate the rates that follow from the current primary variables.
-        std::vector<Scalar> well_q_s = computeCurrentWellRates(simulator, deferred_logger);
+        // Calculate rates at bhp limit, or 1 bar if no limit.
+        std::vector<Scalar> well_q_s(this->number_of_phases_, 0.0);
+        bool rates_evaluated_at_1bar = false;
+        {
+            const auto& summary_state = simulator.vanguard().summaryState();
+            const auto& prod_controls = this->well_ecl_.productionControls(summary_state);
+            const double bhp_limit = std::max(prod_controls.bhp_limit, 1.0 * unit::barsa);
+            this->computeWellRatesWithBhp(simulator, bhp_limit, well_q_s, deferred_logger);
+            // Remember of we evaluated the rates at (approx.) 1 bar or not.
+            rates_evaluated_at_1bar = (bhp_limit < 1.1 * unit::barsa);
+            // Check that no rates are positive.
+            if (std::any_of(well_q_s.begin(), well_q_s.end(), [](Scalar q) { return q > 0.0; })) {
+                // Did we evaluate at 1 bar? If not, then we can try again at 1 bar.
+                if (!rates_evaluated_at_1bar) {
+                    this->computeWellRatesWithBhp(simulator, 1.0 * unit::barsa, well_q_s, deferred_logger);
+                    rates_evaluated_at_1bar = true;
+                }
+                // At this point we can only set the wrong-direction (if any) values to zero.
+                for (auto& q : well_q_s) {
+                    q = std::min(q, Scalar{0.0});
+                }
+            }
+        }
 
         if (nonzero_rate_index == -1) {
-            // No nonzero rates.
-            // Use the computed rate directly
+            // No nonzero rates on input.
+            // Use the computed rate directly, or scaled by a factor
+            // 0.5 (to avoid too high values) if it was evaluated at 1 bar.
+            const Scalar factor = rates_evaluated_at_1bar ? 0.5 : 1.0;
             for (int p = 0; p < this->number_of_phases_; ++p) {
-               ws.surface_rates[p] = well_q_s[this->flowPhaseToModelCompIdx(p)];
+               ws.surface_rates[p] = factor * well_q_s[p];
             }
             return;
         }
 
-        // Set the currently-zero phase flows to be nonzero in proportion to well_q_s.
+        // If we are here, we had a single nonzero rate for the well,
+        // typically from a rate constraint. We must make sure it is
+        // respected, so if it was lower than the calculated rate for
+        // the same phase we scale all rates to match.
         const Scalar initial_nonzero_rate = ws.surface_rates[nonzero_rate_index];
-        const int comp_idx_nz = this->flowPhaseToModelCompIdx(nonzero_rate_index);
-        if (std::abs(well_q_s[comp_idx_nz]) > floating_point_error_epsilon) {
+        const Scalar computed_rate = well_q_s[nonzero_rate_index];
+        if (std::abs(initial_nonzero_rate) < std::abs(computed_rate)) {
```

Comment:
Why do we need to check `std::abs(initial_nonzero_rate) < std::abs(computed_rate)`? It might save some small computation potentially, while it add more complication to the code.  The results should be the same?

Reply:
If `initial_nonzero_rate` is greater than `computed_rate` we should *not* scale everything to match that rate, because the computed_rate comes from a bhp limit and we should not go above that rate (i.e. `factor` would be > 1).

An example: we have a GRAT control set to a high number as an important constraint for the later parts of the simulation, but at the start of simulation the gas rate is low. Then we might scale the rate up to be much higher than the BHP-limit rate.